service/tablet_allocator.cc

```diff
@@ -1016,6 +1052,22 @@ class load_balancer {
                 auto src = ret->second;
                 auto dst = ret->first;
 
+                utils::get_local_injector().inject("forbid_cross_rack_migration_attempt", [&] {
+                    if (cross_rack_migration(src, dst)) {
+                        on_fatal_internal_error(lblogger, "Cross rack colocation is not allowed, killing the node");
+                    }
+                });
+                // If rack awareness is requested, cross-rack migrations cannot be emitted. This will only happen if
+                // balancer wasn't able to achieve rack balance within a replica set, e.g. starting with single rack
+                // and later changing topology to achieve RF=#racks, but this is a rare scenario we don't support now,
+                // so co-location of this table can be postponed indefinitely. Later, we can support this case by
+                // patching balancer to detect rack imbalance in a replica set and fixing it.
+                if (cross_rack_migration(src, dst)) {
+                    lblogger.debug("Cannot emit co-location migration across racks for {}@{} (rack: {}) to co-habit {}@{} (rack: {})",
```

Comment:
warn level seems more appropriate.

Reply:
I am afraid this could spam the log, since it will be dumped every time balancer wakes up (60s interval). The constraint RF != #racks is not in place yet.

Reply:
If merging gets stuck due to this, the reason better be visible.  printing every 60 seconds is not too bad.